Implemented legacy-compatible support for disabling file logging

adamkewley opened the pull request:
- Added a compiler flag, OPENSIM_DISABLE_LOG_FILE, which prevents
  the logger from automatically creating "opensim.log"

- Reworked Logger implementation to only initialize the log file
  sink (i.e. 'opensim.log') when:

  - The logger is about to write its first log message to the file.
    This means that 'opensim.log' will only be created if logging
    occurs

  - Logger::removeFileSink() has not been called before the first
    log message is written to the log. This enables developers to
    runtime-disable the log by calling Logger::removeFileSink() in
    their application *before* logging anything. Note: if the
    logger writes a message during static initialization, then the
    file will be created.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/opensim-org/opensim-core/2880)
<!-- Reviewable:end -->

aymanhab commented:
Thanks @adamkewley I expect to see the symbol exposed in CMake so that it gets wired correctly.

adamkewley commented:
Yeah, I'll get on and add it into CMake also!

On Fri, Sep 25, 2020, 6:23 PM Ayman Habib <notifications@github.com> wrote:

> Thanks @adamkewley <https://github.com/adamkewley> I expect to see the
> symbol exposed in CMake so that it gets wired correctly.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/opensim-org/opensim-core/pull/2880#issuecomment-699023679>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABEC5SVQ44BGLOKMQZXS6PDSHS7WBANCNFSM4RZSCECA>
> .
>

adamkewley commented:
Cool, this feature has also been added to CMake and does seem to propagate correctly when I gave it a whirl.

aymanhab commented on OpenSim/Common/Logger.cpp:
```diff
@@ -30,23 +30,93 @@
+// when filesink initialization *may* happen, ensure it only happens exactly
+// once globally by wrapping it in a function-local static.
+static bool _maybeInitFileLogging() {
+#ifdef OPENSIM_DISABLE_LOG_FILE
```
Style: we never use leading underscore. The name maybeInitFileLogging is less informative, how about initFileLoggingAsNeeded, or something along these lines

aymanhab commented on CMakeLists.txt:
```diff
@@ -111,6 +111,24 @@ mark_as_advanced(OPENSIM_PYTHON_STANDALONE) # Mainly for packagers; not users.
+'opensim.log' file in the current working directory. All log messages
+written by OpenSim (incl. during static initialization) are written to
+both this file and the standard output streams.
+
```
Nice explanation, but a bit too long to fit on a popup in CMake.  Can we condense to match other options, second paragraph could suffice.

aymanhab commented:
Overall looks good, thanks @adamkewley Minor suggestions on style and format.

adamkewley commented on OpenSim/Common/Logger.cpp:
```diff
@@ -30,23 +30,93 @@
+
+// default logger will be initialized during static initialization time
+static std::shared_ptr<spdlog::logger> default_logger = []() {
+    std::shared_ptr<spdlog::logger> logger = spdlog::default_logger();
```
Just so I understand, what is the C++11 syntax for initializing a static variable exactly once at static initialization time, bar writing a separate function?:

```c++
T f() {
    // multiple init steps
}

static T = f();
```

(assuming the initialization has extra steps beyond the constructor)

adamkewley replied:
Just to clarify what I'm getting at - the existing way was *definitely not* the way to do it. C++ doesn't have static classes, or static constructors (unlike e.g. Java) and the idea of constructing an instance of the class to initialize static member variables was asking for an issue somewhere.

Entirely unrelated to this PR (purely coincidence), writing static init code like that is exactly was why we ended up with a memory leak:

https://github.com/opensim-org/opensim-core/issues/2865

The only way to guarantee one-time global initialization is by having a single initializing statement for each element being initialized, or to declare a non-static class with non-static members and *then* initialize a static instance of that class:

```c++
T f() { ... }
static T t = f();  // ok

class C { T m1 = f(); T m2 = f(); };
static C t;  // also ok

// same as first case, but don't expose `f` for use anywhere else:
static T t = []() { ... }();
```

aymanhab replied:
Thanks @adamkewley I was advocating for the first syntax despite the fact that it exposes the name unnecessarily  simply because the rest of the code is c++03 (my bad I claimed c++11) and I couldn't find any instance of the unnamed function initialization in our code base . Purely code style.

adamkewley replied:
Ah - got it. I could change it to use named functions if it's anoying enough (it's not a cross i'm willing to die on)

aymanhab replied:
Let's do it then. Big thanks for accommodating 👍

adamkewley commented:
Cool, I'll clean this up and ship it - thanks @aymanhab

adamkewley commented:
Ok @aymanhab I've applied your suggested cleanups: thank you again.

I also added a note into the CHANGELOG and updated the implementation to set `filesink_auto_initialization_disabled` to `true` more aggressively than previous. Probably not a real issue, but I imagined a sequence such as:

- Application boots, no filesink allocated
- User calls `addFileSink` with their own file log
- Message is logged, triggering auto-initialization
- Warning printed to output along `Already logging to {manual_file}`

Not a hard-error, but logically incorrect: calling `addFileSink` either automatically or manually should just disable the lazy initialization of the log. The other edge-case sequence is:

- User adds a custom log
- User removes a custom log
- First message logged
- Auto initializer creates a log

That's particularly niche, but it would be surprising if it ever happened (the user's probably expecting no log at all). So the new implementation effectively says "as soon as `addFileSink` or `removeFileSink` are called, disable auto-initialization of `opensim.log`" (note: the auto-initializer is a function-local static, so it would never double-call)

adamkewley commented:
I think this should be good to go @aymanhab

adamkewley commented:
Going to go ahead and merge this: the last couple of post-review commits were mostly focused on comments and changing from lambdas, so I'm assuming this is safe to merge now.
